util/unix_util/non_password_auth_user.go

```diff
@@ -0,0 +1,47 @@
+//go:build unix && (!linux || (!amd64 && !arm64))
```

Comment:
I'm not following what this is saying. Are you trying to build on !linux? What's the architecture checks adding here?

Reply:
That should be removed at some point,  I added these checks to make the release script that does cross-compilation (`.github/workflows/build-release.yml`) work correctly as currently this script cannot cross-compile for anything else than linux arm64.
These checks are present so that we use `non_password_auth_user.go` for macos, bsd and linux other than amd64 and arm64. This file allows compiling the server without the CGO code used for password-based auth. That basically provides a server with password-based auth disabled.

For linux amd64 and arm64, `util/unix_util/linux_user.go` is used instead. I agree that every linux should compile correctly using `util/unix_util/linux_user.go` when compiled natively, but that would require a correct cross-compilation toolchain in `.github/workflows/build-release.yml` to make it work when building releases for non-amd64/arm64 archs.

So I know, right now the code is a bit tied to the release-building script until we manage to get rid of CGO. But maybe you know a cleaner way do make that work ?

Reply:
Interesting question there. I think the issue you're trying to solve is how to build on platforms without the necessary headers or os features available in cicd. I think the build flag approach is a good idea here. You could rely on platform specific installers to solve the edge cases like apt, homebrew. 

I didn't see what functionality was offered with `cgo linux`. Is it just password auth (pam)? That's not a super interesting feature for `ssh`, so we could offer binaries without it. If you want `pam` then build from source `go install ...`

Reply:
It is just password auth, yes. Right now, it's using `/etc/shadow` through glibc's `getpwnam`, and not PAM.

PAM would still be cool to have at it provides some features such as motd or even 2-factor auth. There are packages such as [this one](https://pkg.go.dev/github.com/google/fscrypt/pam) that looks promising but it relies on `libpam.so`.

> That's not a super interesting feature for ssh, so we could offer binaries without it. If you want pam then build from source go install ...

Yes, right now we have this [pre-release](https://github.com/francoismichel/ssh3/releases/tag/v0.1.4-bsd) that ships a server without password auth for macos, openbsd, freebsd and linux i386 and armv6.
I still think would be cool to have PAM at some point though but that would mean we can't get rid of CGO (at least in our dependencies).

Reply:
I've added the `disable_password_auth` build tag that can be used on Linux when cross-compiling without full CGO support.  (cf 5cdd68b3b3c486d7c4416c8bf71b752230d17957)